--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -3000,6 +3000,9 @@ static int hostapd_ctrl_iface_req_beacon
 static int hostapd_ctrl_iface_show_neighbor(struct hostapd_data *hapd,
 					    char *buf, size_t buflen)
 {
+    struct hostapd_iface *iface = hapd->iface;
+    struct hostapd_sta_info *info;
+    struct os_reltime now;
 	if (!(hapd->conf->radio_measurements[0] &
 	      WLAN_RRM_CAPS_NEIGHBOR_REPORT)) {
 		wpa_printf(MSG_ERROR,
