Fixed parameter name typos - gs (get-set) to g (get)
--- a/component_dev/tools/linux/cfg80211_ven_cmd.h
+++ b/component_dev/tools/linux/cfg80211_ven_cmd.h
@@ -2119,9 +2119,9 @@ struct vendor_commands vap_vendor_cmds[]
 	{"scsainteropphy",      IEEE80211_PARAM_CSA_INTEROP_PHY, SET_PARAM, 1},
 	{"gcsainteropphy",      IEEE80211_PARAM_CSA_INTEROP_PHY, GET_PARAM, 0},
 	{"scsainteropbss",      IEEE80211_PARAM_CSA_INTEROP_BSS, SET_PARAM, 1},
-	{"gscsainteropbss",     IEEE80211_PARAM_CSA_INTEROP_BSS, GET_PARAM, 0},
+	{"gcsainteropbss",     IEEE80211_PARAM_CSA_INTEROP_BSS, GET_PARAM, 0},
 	{"scsainteropauth",     IEEE80211_PARAM_CSA_INTEROP_AUTH, SET_PARAM, 1},
-	{"gscsainteropauth",    IEEE80211_PARAM_CSA_INTEROP_AUTH, GET_PARAM, 0},
+	{"gcsainteropauth",    IEEE80211_PARAM_CSA_INTEROP_AUTH, GET_PARAM, 0},
 	{"he_amsdu_in_ampdu_supp",      IEEE80211_PARAM_HE_AMSDU_IN_AMPDU_SUPRT,
 		SET_PARAM, 1},
 	{"get_he_amsdu_in_ampdu_supp",  IEEE80211_PARAM_HE_AMSDU_IN_AMPDU_SUPRT,
